src/rgw/rgw_auth_s3.cc

```diff
@@ -538,6 +538,12 @@ int parse_v4_credentials(const req_info& info,                     /* in */
   credential_scope = credential.substr(pos + 1);
   ldpp_dout(dpp, 10) << "credential scope = " << credential_scope << dendl;
 
+  std::string x_amz_content_sha256 = std::string(info.env->get("HTTP_X_AMZ_CONTENT_SHA256", ""));
+  std::regex hexRegex("^[0-9a-fA-F]+$");
+  if (x_amz_content_sha256.length() != 64 || !std::regex_match(x_amz_content_sha256, hexRegex)) {
+    ldpp_dout(dpp, 10) << "error x-amz-content-sha256 = " << x_amz_content_sha256 << dendl;
+    return -EINVAL;
```

Comment:
don't we have to support all of the special values like UNSIGNED-PAYLOAD etc here? https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-auth-using-authorization-header.html#w33934aab9c25b9c32

Reply:
It seems to support the validation of the x-amz-content-sha256 request header in the bool AWSv4ComplSingle::complete() function in the rgw_auth_s3.cc file. The validation of the header and the signature is done separately. However, why is it possible to create a bucket successfully even when x-amz-content-sha256 is incorrect?

Reply:
if the CreateBucket request body is empty, we use a null completer instead of AWSv4ComplSingle. see `is_v4_payload_empty()` in:
https://github.com/ceph/ceph/blob/d7e2a322a5490d1fc9395471e25c3a98499975ea/src/rgw/rgw_rest_s3.cc#L5820-L5827

if the request body isn't empty (for example, from `s3cmd mb s3://testbucket --bucket-location=foo` which sends a CreateBucketConfiguration xml document), i do see `AWSv4ComplSingle` being created and later checked in do_aws4_auth_completion:
```
CONTENT_LENGTH=102
...
req 9584837891480720678 0.000000000s delaying v4 auth
...
req 9584837891480720678 0.001000034s s3:create_bucket NOTICE: call to do_aws4_auth_completion
req 9584837891480720678 0.001000034s s3:create_bucket v4 auth ok -- do_aws4_auth_completion
req 9584837891480720678 0.001000034s s3:create_bucket create bucket input data=<CreateBucketConfiguration><LocationConstraint>foo</LocationConstraint></CreateBucketConfiguration>
```

Reply:
since is_v4_payload_empty() is the only case we're missing, i pushed a draft commit to https://github.com/ceph/ceph/pull/55250 that adds it right before that null-completer part. how does that look?

cc @mattbenjamin who's working on other completer changes in https://github.com/ceph/ceph/pull/54856

Reply:
OK, it is also a good idea.